### ISSUE TITLE
Appearance/Disappearance of a scene?

### ISSUE THREAD

**Issue by emerson233** [grahammendick/navigation#321]
How do we know when a scene is going to appear/disappear? In native iOS development it's very common to do some extra work in `viewWillAppear`/`viewDidAppear`/`viewWillDisappear`/`viewDidDisappear`. Is it supported in this repo currently?

**Comment by grahammendick** [comment]
Can you explain what you're trying to do, please? Then I can advise you the best way to accomplish it with the Navigation router.

**Comment by emerson233** [comment]
A simple user case:
We have a page contains a camera view. We want to open the camera after the page appears fully on the screen to avoid UI lags.(Opening camera is an expensive operation.)

**Comment by grahammendick** [comment]
I'm not convinced this is a realistic example. Why wouldn't you open the camera directly when the user taps a button instead of navigating to a new scene?

**Comment by emerson233** [comment]
Well, I think it's pretty reasonable to open a camera to scan a QR Code in a new scene. You can check [this similar feature](https://faq.whatsapp.com/en/web/28080003/) in WhatsApp.

**Comment by grahammendick** [comment]
So it does! Thanks for that example.
You can delay the camera opening by setting a flag in the componentDidMount/useEffect. 
```js
componentDidMount() {
  setTimeout(() => {
    this.setState({showCamera: true});
  }, 200);
}
```

**Comment by emerson233** [comment]
What if we want to do something when a scene appears from calling `navigateBack`.
Say we have a `MapView` in scene A, we want to keep memory usage as low as possible, so we hide `MapView` after we navigating to scene B and show `MapView` when navigating back to scene A.

**Comment by grahammendick** [comment]
You shouldn't need to hide the `MapVew` because, when you navigate to scene B, Fragment A is paused and the `MapView` detached from the window.

**Comment by emerson233** [comment]
Creating a new `UIViewController` doesn't mean the memory allocated in previous `UIViewController` is freed I think.

**Comment by grahammendick** [comment]
Does iOS recommend to hide the `MKMapView` in `viewDidDisappear` and show it again in `viewWillAppear`?

**Comment by emerson233** [comment]
https://docs.mapbox.com/ios/api/navigation/0.15.0/reduce-memory-usage.html
We are using Mapbox.

**Comment by grahammendick** [comment]
Can you explain a bit more about how you integrate with Mapbox? The link you sent talks about their own custom  [NavigationViewController](https://docs.mapbox.com/ios/api/navigation/0.15.0/Classes/NavigationViewController.html) that inherits from `UIViewController`. How are you showing that?

**Comment by emerson233** [comment]
Sorry, it's [react-native-mapbox](https://github.com/react-native-mapbox-gl/maps) we are using.
[This reference](https://docs.mapbox.com/ios/api/navigation/0.15.0/reduce-memory-usage.html) illustrates that it's indeed necessary to remove `MapView` after navigating to other `ViewController`(including `NavigationViewController`). The point is `NavigationController` contains a `MapView` also. We don't want to keep two `MapView` instances in memory.

**Comment by emerson233** [comment]
Current implementation using react-native-navigation:  render MapView when receiving `willAppear` event.
```javascript
render() {
  return (this.state.appeared ? <MapView /> : null)
}
```

**Comment by grahammendick** [comment]
I'm still don't think this is necessary. Your [MapBox reference](https://docs.mapbox.com/ios/api/navigation/0.15.0/reduce-memory-usage.html) is talking about hiding a preview map when **presenting** their NavigationViewController. Presenting means it's showing in a modal and so the previous ViewController is still active.

When you navigate to a new scene, it pushes the new ViewController and so the previous ViewController isn't active anymore.

**Comment by emerson233** [comment]
I don't see the difference between `pushing` and `presenting` a view controller here.

> When you navigate to a new scene, it pushes the new ViewController and so the previous ViewController isn't active anymore.

What do you mean by saying the previous ViewController isn't active? Doesn't it still stay in NavigationStack?

**Comment by grahammendick** [comment]
If you click the 'Debug View Hierarchy' button in Xcode and you'll see the difference. When pushed, the old ViewController isn't in the hierarchy. When presented the old ViewController is still there.

**Comment by emerson233** [comment]
Yes. It's not in the debug view hierarchy, but it is consuming memory anyway. I believe it's still active.

**Comment by grahammendick** [comment]
iOS already optimizes the memory for inactive screens, so I don't think you need to optimize it further. MapBox doesn't say anything about hiding it in `viewDidDisappear` and showing it in `viewWillAppear`.

**Comment by emerson233** [comment]
> iOS already optimizes the memory for inactive screens

Could you post a reference linking to this one please?

**Comment by grahammendick** [comment]
[React Navigation explained the idea nicely](https://reactnavigation.org/docs/en/react-native-screens.html)

**Comment by emerson233** [comment]
Well, I don't think the optimization is good enough to reduce the `MapView` memory usage. Tested it out a moment ago, `MapView` is consuming massive memory (100+MB) even it's "inactive".
In this [related isssue](https://github.com/mapbox/mapbox-gl-native/issues/9203), the author is using `UINavigationController` and still have a problem.

**Comment by grahammendick** [comment]
The [related issue](https://github.com/mapbox/mapbox-gl-native/issues/9203) is only about a memory problem when there are two scenes each with a `MapView`. It doesn't suggest doing anything if there is only one `MapView`.

If we navigate from A --> B --> C where C has a MapView then the memory footprint might be 
1MB + 1MB + 100MB = 102MB. Assuming that scenes without a MapView have memory consumption of 1MB each.  If you then navigate to D, the memory footprint only goes up to 103MB. So if we have a memory problem on scene D then we would've had a memory problem on scene C. If we need to hide the MapView on scene D then we'd need to hide it on scene C too. But we can't hide it on scene C because that's where the MapView is.

**Comment by emerson233** [comment]
Sorry I didn't describe my question correctly.

> Tested it out a moment ago, MapView was consuming massive memory (100+MB) even it was "inactive".

I mean we tried to navigation from A --> B --> C where both B and C contain a MapView, and when we navigated to scene C finally, the memory went up to 200+MB. Assuming each MapView used 100MB memory, we intended to free the memory allocated in scene B, and memory usage should go down to 100+MB.

**Comment by grahammendick** [comment]
Cool, thanks for explaining your scenario in more detail. Can you explain a bit more about how you handled it with react native navigation, please? You said you listened for the 'willAppear' event but I don't see that event listed in their [events list](https://wix.github.io/react-native-navigation/#/docs/events). And I found an issue that says [these events were dropped](https://github.com/wix/react-native-navigation/issues/4793#issuecomment-467748525) because they're not reliable.

**Comment by emerson233** [comment]
hmm...We are using react-native-navigation v1 so...
Anyway, could it be possible to emit `didAppear/didDisappear` event in this repo?
Or if you are still not convinced to add this feature, could you tell me what's the best approach to get this done?

**Comment by grahammendick** [comment]
There already are `didAppear` and `didDisappear` events in the Navigation router. You listen for navigation events and check if the current `crumb` matches your `crumb`. The `crumb` is the index of the scene.
```js
const crumb = stateNavigator.stateContext.crumb;
stateNavigator.onNavigate((oldState, state, data, _, stateContext) => {
  if (stateContext.crumb === crumb)
    // did Appear
  else
    // did Disappear
})
```
I'm not sure this will work for your scenario though because if the user swipes back slowly then the `MapView` doesn't appear until the swipe completes.

**Comment by emerson233** [comment]
Thanks! Will give it a try.
Close this issue for now.

**Comment by grahammendick** [comment]
Don't forget to remove your listeners, too
```js
handleNavigation(oldState, state, data, _, stateContext) {
  const crumb = stateNavigator.stateContext.crumb;
  if (stateContext.crumb === crumb)
    // did Appear
  else
    // did Disappear
})
componentDidMount() {
  this.stateNavigator.onNavigate(this.handleNavigation);
}
componentWillUnmount() {
  this.stateNavigator.offNavigate(this.handleNavigation);
}
```

**Comment by grahammendick** [comment]
Hey @emerson233 . How's it going?

I've just released `navigation-react-native v6.15.0`. It supports [scene navigation lifecycle hooks](https://grahammendick.github.io/navigation/documentation/native/scene-navigation-lifecycle.html). In your Map example, you can add hooks to scene B and show and hide the map when the user navigates to and from B
```js
useNavigated(() => {
  // Show map
});
useUnloaded(() => {
  // Hide map
});
```
